text/2018-10-25-tikv-client-rust.md

```diff
@@ -0,0 +1,386 @@
+
+The client provides two types of APIs in two separate modules for developers to choose from.
+
+### The common data types
+
+- Key: raw binary data
+- Value: raw binary data
+- KvPair: Key-value pair type
+- KeyRange: Half-open interval of keys
+- Config: Configuration for client
+
+### Try the Raw Key-Value API
+
+To use the Raw Key-Value API, take the following steps:
+
+1. Create an instance of Config to specify endpoints of PD (Placement Driver) and optional security config
+
+```rust
+    let config = Config::new(vec!["127.0.0.1:2379"]);
+```
+
+2. Create a Raw Key-Value client.
+
+```rust
+    let client = RawClient::new(&config);
+```
+
+3. Call the Raw Key-Value client methods to access the data on TiKV. The Raw Key-Value API contains following methods
+
+```rust
+    fn get<K, C>(&self, key: K, cf: C) -> KvFuture<Value>
+    where
+        K: Into<Key>,
+        C: Into<Option<String>>;
+
+    fn batch_get<I, K, C>(&self, keys: I, cf: C) -> KvFuture<Vec<KvPair>>
+    where
+        I: IntoIterator<Item = K>,
+        K: Into<Key>,
+        C: Into<Option<String>>;
+
+    fn put<P, C>(&self, pair: P, cf: C) -> KvFuture<()>
+    where
+        P: Into<KvPair>,
+        C: Into<Option<String>>;
+
+    fn batch_put<I, P, C>(&self, pairs: I, cf: C) -> KvFuture<()>
+    where
+        I: IntoIterator<Item = P>,
+        P: Into<KvPair>,
+        C: Into<Option<String>>;
+
+    fn delete<K, C>(&self, key: K, cf: C) -> KvFuture<()>
+    where
+        K: Into<Key>,
+        C: Into<Option<String>>;
+
+    fn batch_delete<I, K, C>(&self, keys: I, cf: C) -> KvFuture<()>
+    where
+        I: IntoIterator<Item = K>,
+        K: Into<Key>,
+        C: Into<Option<String>>;
+
+    fn scan<R, C>(&self, range: R, limit: u32, key_only: bool, cf: C) -> KvFuture<Vec<KvPair>>
+    where
+        R: Into<KeyRange>,
+        C: Into<Option<String>>;
+
+    fn batch_scan<I, R, C>(
+        &self,
+        ranges: I,
+        each_limit: u32,
+        key_only: bool,
+        cf: C,
+    ) -> KvFuture<Vec<KvPair>>
+    where
+        I: IntoIterator<Item = R>,
+        R: Into<KeyRange>,
+        C: Into<Option<String>>;
+
+    fn delete_range<R, C>(&self, range: R, cf: C) -> KvFuture<()>
+    where
+        R: Into<KeyRange>,
+        C: Into<Option<String>>;
+```
+
+#### Usage example of the Raw Key-Value API
+
+```rust
+extern crate futures;
+extern crate tikv_client;
+
+use futures::future::Future;
+use tikv_client::raw::Client;
+use tikv_client::*;
+
+fn main() {
+    let config = Config::new(vec!["127.0.0.1:3379"]);
+    let raw = raw::RawClient::new(&config)
+        .wait()
+        .expect("Could not connect to tikv");
+
+    let key: Key = b"Company".to_vec().into();
+    let value: Value = b"PingCAP".to_vec().into();
+
+    raw.put((Clone::clone(&key), Clone::clone(&value)), None)
```

Comment:
It would be very nice to avoid the clones here.

Reply:
Hm, you're right. I think in the case of put it's okay... I note that `HashMap::get` takes a borrow though: https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.get

I think in this case mutations, eg `put`, which actually go and send the value (and thus almost certainly need to do memory work) should take. But maybe we can make the searching/get based functions accept just a borrow? Similar to `HashMap` or `HashSet`?

Reply:
The gRPC part is going to move whatever parameters passed in anyway. In addition the nature of asynchronous programming model requires moving and sending parameters to event loop thread, so it doesn't matter if it's mutation or not the arguments are going to be moved anyway. However I do feel this particular API guideline applies to regular libraries better then the scenario we are facing. So breaking this rule probably is a better choice as it makes our users' life easier.

Reply:
I changed it to AsRef<Key> and it does look better. Please take a look at the new commit. @Hoverbear

Reply:
@siddontang What do you think about this?